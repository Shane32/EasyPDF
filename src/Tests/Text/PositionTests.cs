using System.Drawing;
using Font = Shane32.EasyPDF.Font;

namespace Tests.Text;

public class PositionTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public PositionTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Theory]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.18f, 0.03f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1f, 6f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.18f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 6f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.216f, 0.036f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 0f, 1.05f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 1f, 1.027f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 1f, 1.233f)]
    public void TextSizes(float lineSpacing, float paragraphSpacing, float stretchY, float capHeight, float ascent, float descent, float height, float leading, float paragraphLeading, float stretchX, float characterSpacing, float width)
    {
        _writer.Font = new Font(StandardFonts.Times, 12f) {
            LineSpacing = lineSpacing,
            ParagraphSpacing = paragraphSpacing,
            StretchY = stretchY,
            CharacterSpacing = characterSpacing,
            StretchX = stretchX,
        };
        _writer.TextCapHeight().ShouldBe(capHeight, 0.001f);
        _writer.TextAscent().ShouldBe(ascent, 0.001f);
        _writer.TextDescent().ShouldBe(descent, 0.001f);
        _writer.TextHeight().ShouldBe(height, 0.001f);
        _writer.TextLeading().ShouldBe(leading, 0.001f);
        _writer.TextHeight().ShouldBe(_writer.TextAscent() + _writer.TextDescent() + _writer.TextLeading(), 0.001f);
        _writer.TextParagraphLeading().ShouldBe(paragraphLeading, 0.001f);
        _writer.TextWidth("Hello World!").ShouldBe(width, 0.001f);
    }
    
    [Fact]
    public void Standard()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void CharacterSpacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { CharacterSpacing = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
