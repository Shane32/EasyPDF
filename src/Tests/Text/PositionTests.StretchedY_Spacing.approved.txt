%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 921/Filter/FlateDecode>>stream
x\x9C\x95\x97OO\xDB@\x10\xC5\xEF\xFE\x14>\xC2!\xEE\xCE\xDA\xBB\xF6\xF6V\xAA\xB6\x12\xEA\xA5\xD47\xC4\xA1
\x81R\x11(4R\xBF~\xD73;\x7F\x8C\x88\xD8
H&\x9B\x9D\xF1\xCB\x0F\xE7=\xFB\xA99\x9B\x9B>\xB6c\x88\xED|\xDD|\x9A\x9Bo\xB4.\xFFl\xA0\x1D}\xFEu\xEDv\xDF\xB8\xF6<\xFF\xFDj\\\x07\xED\xDF\xE6\xF2*\xBF\x7F\xDD8\xDCw\xF1\xA5\x14\xCF\xB7\xAFo\xF3\xDEu!W\xFBR\xF9nH\xAD\x87iy\x1A\x96\xD7\x10\xF1y\xDBx\xE8\xBB\x00e\x11t+`\xFF\xF2\xFERm\xCA*\xEE\xDD\xF0\x84\x8F\xC0\xB9\xB4\x8C\x87+\xDB\x8B\x88m\xF3\xBD\xE1\x81{\x91v\x8F\xABe\xC8\xBEaE\xCB\xF2S\xFE@\xF3v\xA1\xF4\xEE3\xB4\xA1\x9Do\x98\x8F\xC7\xE3\x8E\x1DL\xED\xBC\xB7\x0C\xDCB\xF2\xF2\xE4\xC3\xFD\xDD\xED\xC3~\xF7px\x7F\x1A\xDCI\xFBuws\x98O\xC1\xBB\x93\xC7\xDF\xA7W\xF39\xB1&\x05y\x84\xD0q\x06\x8F\xF7\xE5\xD3a!\x80\xCA2\x98\xED@S\x98\x11\x04\xDCG\xDB\x97G\x9C\x82\x850\xA2e:>mg-
\xC9\x08\xCB%c\xC2N\xC2\x84Rj8\xF9<.U\x83\xFA\x98_\xED\x9E_b\xCA\xE7)\xAB\xE9'{\x169>\x8D\xDC\xEA<r\x82\x89\xB7\x03MaL}o0\xF5P\xA6`!\x98h\x99\x0E\xDA+&\x9C\xC2\x98\xAC\xB0(\x98\xB0\xB3\x9CM\xAE\x12Sn\xAFet\xF6\xE3\xCF\xEE\xFE\xEEa\xF7\x92R\x18DL\x88\x86R\x98\xCA\xE7\xC3B(\x95e0\xDB\x81\xA60\xA5\x00\x86\xD2\x90\x98u\xB2\x94h\x99\x0E
J\x09\xA70%+l\x10J\xD8I\x94PJ\xA5\xE0:W\xFF\xAD;{<\x1C\x1E\xF7/1esc5\xE3`0\x8D\xB1|@,\x04SY\x06\xB3\x1Dh
c\x8A\xC9`\x8Ac\x99\x82\x85`\xA2e<h\xD9\xCEZ\x14\x93\x15\xE6\x05\x13v\x12&\x94\xF2\x16&H\xBE\x83,;\xD5\xFA\x13}\xE5\x8E9TR\xBBL\xD6\xC0\x13\xFBoZYxR\x0FO\xC6\xC4\x93\xBA\xF84\x1AZS(S\xB0\x10Z\xB4\x8C\x07-\xDBY\x8B\xD2\xB2\xC2\xD4\xC8\xB1\x93h\xA5\x1A'\x87i\xEAb\x96\xDD\xD7\x9A\x14\xD1z\xDD\xA6\xC0\x19\xD7\x04\xEB\xE7\x00l\xE8T\x09/~\x03l\x0B\x94Yb\xEAn\xE5\xEANl\xDD\xAD}\xDD\x19cw\xD6\xD9\xDD\xCA\xDA\xADJg\xCC\xDD\xA9\xBB\x93\xAC\xB7\xE1\xE5\x13\xB2\xC7!\xF5\xE8\x8E\xB9\x17x\xF5R\xF0\xD6\xE5\xA1g\x9B\xA7J\xE1\xF5j\xF4\xD2\x02e\x96\xC0\xF3\xD6\xEB\xC1\xB3\xD9S\xA5\xF0\xBC\xDA=\xB7\x88.\x03o\xA5R-\x9F\xDA\x0B\xBC\xBE\xC6\xF4\x973/?\xE6\x1C\xFD?|\xAFz\x1A\x0Cj\xB10X\xF3\x87\x81\xDD\x9F*\x857\xA8\xFFK\x0B\x94Y\x02o\x80\xD5\xF5D\x92\x7F\xC4*\x04\xCA\x1B\xE5\xF0&\x06h\x96\xC0[\xA9\x1C\xCCe\x85F\x01\xC9z\x13\xDE\x98\xE7-\x97/P\xEBr\x17w\xB7?\x8F_\x86E\xB5^\x886\x14 r*P\xA5\xF8\xA2\xE6\x82\xB4@\x99\xA5Wc6\x1A p6P\xA5\xF8\x82\xA6\x03\xB7\x88.\x83o\xA5R\x13\x82\xDA\x0B\xBEX\x95\x11Yp\x18\xF3S\xA8\xB5=\xC4w\xC4\xF5&g\xAE\x15mF\xC0\xC4!A\x95\xC2\x9B4&\xA4\x05\xCA,\x817\xDA\xA4\x80\x91\xA3\x82*\x857jXp\x8B\xE8\xB2\x17\xB4V\xA5\x06\x06\xB5\x17xSUdD\xD7\xF5\x09\x87T\x93;jz\xC9\xDE\x00\xAC\xEE\x00\x9C\xDC\x02\xB8\xF5=\x80\xB37\x0161\x92I\x8C\xB4J\x8C$\x89\x91\xD6\x89\x91Lb$\x9B\x18i\x95\x18F\xE5R\x0B\xBBd\xEE\x07\\Ub\xC4\xBE\x1Bsb\xE4l\xAEt=\xC2\xF7\xC2\xF4\xFE\x01\xD7\x98\x01\xFB
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001160 00000 n 
0000000015 00000 n 
0000001248 00000 n 
0000001003 00000 n 
0000001314 00000 n 
0000001359 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 